### PR TITLE
Feature/343/sonar source analyzer

### LINKED ISSUE #343: Add Sonar Java to SourceCodeParser
Add Sonar Java Plugin code to the SourceCodeparser that analyzes the metrics 

* [x] complexity 
* [x] functions
* [x] comment_lines
* [x] cognitive_complexity
* [x] ncloc
* [x] classes
* [x] complexity_in_classes
* [x] statements

The Sourcecodeparser should still return the current metrics lines of code and rloc. Mcc should be exactly the same as complexity. If it is, the Mcc metric can be removed.

-----
## Tech
* Please replace the current sourcecodeparser Implementation.

### PR DESCRIPTION
**Pull request by alexhunziker**
# SonarJava for SourceCodeParser

closes #343 
Merge permission: CC-Member

## Description

- Changed structure of SourceCodeParser in order to being able to integrate Sonar plugins for scanning
- Added SonarJava to Source code parser
- Replaced former implementation of former SourceCodeParser completely (including custom metrics)

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

* [x] **All** requirements mentioned in the issue are implemented
* [x] Does match the Code of Conduct and the Contribution file 
* [x] Task has its own **GitHub issue** (something it is solving)
  * [x] Issue number is **included in the commit messages** for traceability
* [x] **Update the README.md** with any changes/additions made
* [x] **Update the CHANGELOG.md** with any changes/additions made
* [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
* [x] **All tests pass**    
* [x] **Descriptive pull request text**, answering:
  + What problem/issue are you fixing?
  + What does this PR implement and how? 
* [x] **Assign your PR to someone** for a code review
  + This person _will be contacted **first**_ if a bug is introduced into `master`
* [x] **Manual testing** did not fail

### CODE REVIEW

**Comment by alschmut** [comment]
I got this error when analyzing this OpenSource [Java project](https://github.com/TheAlgorithms/Java). Do you know anything about it? But nevertheless I got a nice cc.json with 158 leaves.
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by net.sf.cglib.core.ReflectUtils$1 (file:/Users/.../codecharta/analysis/build/install/codecharta-analysis/lib/ccsh-1.24.0.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of net.sf.cglib.core.ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

File: analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/SourceCodeParserMain.kt [hunk 1]
```diff
@@ -1,68 +1,63 @@
...
 import picocli.CommandLine.*
 import java.io.*
 import java.nio.file.Paths
 import java.util.concurrent.Callable
 
-@Command(name = "sourcecodeparser", description = ["generates cc.JSON from source code"],
-        footer = ["Copyright(c) 2018, MaibornWolff GmbH"])
-class SourceCodeParserMain(private val outputStream: PrintStream): Callable<Void> {
-
+@Command(name = "sourcecodeparser", description = ["generates cc.JSON from source code"], footer = ["Copyright(c) 2018, MaibornWolff GmbH"])
```
[the remark this answers is not in the document]
Comment (alexhunziker) [reply analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/SourceCodeParserMain.kt]: ok i will change that, lowercase looks better

**Comment by alexhunziker** [comment]
> WARNING: An illegal reflective access operation has occurred

do you get this error message when executing the built ccsh in a terminal window or just within intelliJ? because i have this message too, but only in intelliJ. when starting from gitbash or the windows terminal this does not happen.

The reason for this message is: sonar-java is using cglib (https://github.com/SonarSource/sonar-java/blob/master/sonar-java-plugin/pom.xml). The issue is within cglib, and it has already been reported. However cglib is no longer developed actively (see https://github.com/cglib/cglib/issues/129). So as far as i can see there is not much we can do here.

If this appears in your terminal if you start it like a user would, we should somehow suppress this warning however.

**Comment by alschmut** [comment]
@alexhunziker no unfortunately I got this warning when I executed it in the macOS bash terminal as a user...

**Comment by alexhunziker** [comment]
Ok, thanks, i will see if we can somehow supress this then.

**Comment by alexhunziker** [comment]
@alschmut can you please check if you still get the warning on your system? I think it should be fixed

**Comment by alexhunziker** [comment]
So apparently we need to set this option somehow in order to have java shut up about the reflection of cglib:
--add-opens java.base/java.lang=ALL-UNNAMED
but i have no idea on how to do this

**Comment by surfmuggle** [comment]
> do you get this error message when executing the built ccsh in a terminal window

Yes i am getting this error using windows 10 and [sonar-scanner-cli-4.2.0.1873-windows.zip](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner/) to scan code to upload it to SonarQube™ for Bitbucket Server

**Pull request merged**